### PR TITLE
metaanalyse

### PR DESCRIPTION
**Pull request by akaTwiiX**
# Description
Add metaanalyse example CMD line program with a little documentation in the README.md about the code.

### CODE REVIEW

File: module-4/metaanalyse/README.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+- csv
+- os
+
+## Installation
+
+Install the required dependencies using pip:
+
+```bash
+pip install PyPDF2
+```
```
Comment (spmse) [inline module-4/metaanalyse/README.md]: **for the requirements:** add a `requirements.txt` and adjust the install command to install requirements from the file.

**General:**
Add an overview about requirements and prerequisites, like e.g. a virtual environment and how to create one locally in order to isolate the package installations

File: module-4/metaanalyse/README.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+```bash
+python3 pdf_metadata_extractor.py -d /path/to/directory -n output.csv
+```
+
+## Arguments
+
+- `-f` or `--file`: Path to a single PDF file.
+- `-d` or `--directory`: Path to a directory containing multiple PDF files.
+- `-n` or `--name`: Name and path of the output CSV file (required).
```
Comment (spmse) [inline module-4/metaanalyse/README.md]: You should use a table to describe this section and add an indicator whether a parameter is required or not.
this section should be placed before `single file extraction` and `multi file extraction` hints.

File: module-4/metaanalyse/README.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+  - `file_path` (str): Path to the PDF file.
+  - `csv_path` (str): Path to the output CSV file.
+
+`process_directory(directory_path, csv_path)`
+
+This function processes multiple PDF files in a given directory, extracts their metadata, and writes it to a CSV file.
+
+- `Parameters`:
+  - `directory_path` (str): Path to the directory containing PDF files.
+  - `csv_path` (str): Path to the output CSV file.
```
Comment (spmse) [inline module-4/metaanalyse/README.md]: as for other submissions this section might be restructured or replaced with links to your code, when the code contains docstrings

File: module-4/metaanalyse/metaanalyse.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+        metadata['Title'] = info.get('/Title', 'undefined') or 'undefined'
+        metadata['Author'] = info.get('/Author', 'undefined') or 'undefined'
+        metadata['Creator'] = info.get('/Creator', 'undefined') or 'undefined'
+        metadata['Created'] = info.get('/CreationDate', 'undefined') or 'undefined'
+        metadata['Modified'] = info.get('/ModDate', 'undefined') or 'undefined'
+        metadata['Subject'] = info.get('/Subject', 'undefined') or 'undefined'
+        metadata['Keywords'] = info.get('/Keywords', 'undefined') or 'undefined'
+        metadata['Description'] = info.get('/Description', 'undefined') or 'undefined'
+        metadata['Producer'] = info.get('/Producer', 'undefined') or 'undefined'
```
Comment (spmse) [inline module-4/metaanalyse/metaanalyse.py]: the second param of the `get` function is optional and will only be returned if the `key` was not found in your dictionary --> the `or 'undefined'` part is unnecessary.

File: module-4/metaanalyse/metaanalyse.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+                file_path = os.path.join(root, file)
+                metadata = extract_pdf_metadata(file_path)
+                metadata_list.append(metadata)
+
+    with open(csv_path, 'w', newline='', encoding='utf-8') as csv_file:
+        fieldnames = ['File Name', 'Title', 'Author', 'Creator', 'Created', 'Modified', 'Subject', 'Keywords', 'Description', 'Producer', 'PDF Version']
+        writer = csv.DictWriter(csv_file, fieldnames=fieldnames, delimiter=';')
+        writer.writeheader()
+        for metadata in metadata_list:
+            writer.writerow(metadata)
```
Comment (spmse) [inline module-4/metaanalyse/metaanalyse.py:62]: this is more or less the same code than above for single file processing -> extract the code into a reusable function that can create your result file for one or more files where metadata should be extracted

File: module-4/metaanalyse/metaanalyse.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    parser.add_argument('-n', '--name', required=True, help='Name and path of the output CSV file.')
+
+    args = parser.parse_args()
+
+    if args.file:
+        process_single_file(args.file, args.name)
+    elif args.directory:
+        process_directory(args.directory, args.name)
+    else:
+        print('Please specify either a file (-f) or a directory (-d).')
```
Comment (spmse) [inline module-4/metaanalyse/metaanalyse.py]: i would recommend turning around the logic to check for:

```python
if not args.file or not args.directory:
    raise Error("You need to specify either -f or -d option")
# handling like above
if args.file:
    process_single_file()
else:
    process_directory()
```

**Review by spmse — CHANGES_REQUESTED** [review]
Hello John! I have written a small code review for you with some small change requests for you to adjust in your code.
In case of questions feel free to ask and if you think you are done re-submit your PR for review :)

Make sure to add some docstrings and corresponding links from docs to code while adjusting your submission

Best regards, Sven

File: module-4/metaanalyse/README.md (hunk 1 above)
Comment (akaTwiiX) [reply module-4/metaanalyse/README.md]: not for 2 items :D

File: module-4/metaanalyse/metaanalyse.py (hunk 6 above)
Comment (akaTwiiX) [reply module-4/metaanalyse/metaanalyse.py]: Exeption :D

File: module-4/metaanalyse/metaanalyse.py [hunk 7]
```diff
@@ -70,12 +70,13 @@ def main():
 
     args = parser.parse_args()
 
+    if not args.file or not args.directory:
```
Comment (spmse) [inline module-4/metaanalyse/metaanalyse.py]: must be `and` - otherwise if you specify `-d` but not `-f` you will result in an error.

**Review by spmse — CHANGES_REQUESTED** [review]
Hello John, I've spotted a small logic error, but if that is fixed you may update the branch and merge it :) (last part depends on your general merge order)

File: module-4/metaanalyse/metaanalyse.py (hunk 7 above)
Comment (akaTwiiX) [reply module-4/metaanalyse/metaanalyse.py]: absolut xD Thx <3

**Pull request merged**